WordPress/Classes/ViewRelated/Post/PostEditorState.swift

```diff
@@ -0,0 +1,292 @@
+        case .save:
+            return NSLocalizedString("Save", comment: "Save button label (saving content, ex: Post, Page, Comment).")
+        case .schedule:
+            return NSLocalizedString("Schedule", comment: "Schedule button, this is what the Publish button changes to in the Post Editor if the post has been scheduled for posting later.")
+        case .submitForReview:
+            return NSLocalizedString("Submit for Review", comment: "Submit for review button label (saving content, ex: Post, Page, Comment).")
+        case .update:
+            return NSLocalizedString("Update", comment: "Update button label (saving content, ex: Post, Page, Comment).")
+        }
+    }
+
+    var publishingActionLabel: String {
+        switch self {
+        case .publish:
+            return NSLocalizedString("Publishing...", comment: "Text displayed in HUD while a post is being published.")
+        case .save:
+            return NSLocalizedString("Saving...", comment: "Text displayed in HUD while a post is being saved as a draft.")
+        case .schedule:
+            return NSLocalizedString("Scheduling...", comment: "Text displayed in HUD while a post is being scheduled to be published.")
+        case .submitForReview:
+            return NSLocalizedString("Submitting for Review...", comment: "Text displayed in HUD while a post is being submitted for review.")
+        case .update:
+            return NSLocalizedString("Updating...", comment: "Text displayed in HUD while a draft or scheduled post is being updated.")
+        }
+    }
+
+    var isPostPostShown: Bool {
+        switch self {
+        case .publish:
+            return true
+        default:
+            return false
+        }
+    }
+}
+
+fileprivate protocol PostEditorActionState {
+    var action: PostEditorAction { get }
+
+    // Actions that change state
+    func updated(postStatus: PostStatus, context: PostEditorStateContext) -> PostEditorActionState
+    func updated(publishDate: Date?, context: PostEditorStateContext) -> PostEditorActionState
+}
+
+public protocol PostEditorStateContextDelegate {
+    func context(_ context: PostEditorStateContext, didChangeAction: PostEditorAction)
+    func context(_ context: PostEditorStateContext, didChangeActionAllowed: Bool)
+}
+
+public class PostEditorStateContext {
+    private var editorState: PostEditorActionState {
+        didSet {
+            delegate?.context(self, didChangeAction: editorState.action)
+        }
+    }
+
+    private var publishActionAllowed = false {
+        didSet {
+            delegate?.context(self, didChangeActionAllowed: publishActionAllowed)
+        }
+    }
+
+    fileprivate var originalPostStatus: PostStatus
+    fileprivate var userCanPublish: Bool
+    private var delegate: PostEditorStateContextDelegate?
+
+    fileprivate var hasContent = false {
+        didSet {
+            publishActionAllowed = hasContent && !isBeingPublished
```

Comment:
I know it's only one line, but we could extract this to a private `updatePublishActionAllowed` func and call it from both `didSet`s?